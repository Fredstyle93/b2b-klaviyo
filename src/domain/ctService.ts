import { AuthMiddlewareOptions, Client, ClientBuilder, HttpMiddlewareOptions } from '@commercetools/sdk-client-v2';
import { ApiRoot, createApiBuilderFromCtpClient, Customer, Order } from '@commercetools/platform-sdk';
import logger from '../utils/log';
import fetch from 'node-fetch';
import { StatusError } from '../types/errors/StatusError';

type ByProjectKeyRequestBuilder = ReturnType<ApiRoot['withProjectKey']>;

let ctClient: Client;

export const createCtClient = () => {
    if (!ctClient) {
        const ctApiCredentials = JSON.parse(process.env.CT_API_CLIENT || '{"clientId":"","secret":""}');

        const authMiddlewareOptions: AuthMiddlewareOptions = {
            host: process.env.CT_AUTH_URL || '',
            projectKey: process.env.CT_PROJECT_ID || '',
            credentials: {
                clientId: ctApiCredentials.clientId || '',
                clientSecret: ctApiCredentials.secret || '',
            },
            scopes: [process.env.CT_SCOPES || ''],
            fetch,
        };

        const httpMiddlewareOptions: HttpMiddlewareOptions = {
            host: process.env.CT_API_URL || '',
            fetch,
        };

        ctClient = new ClientBuilder()
            .withClientCredentialsFlow(authMiddlewareOptions)
            .withHttpMiddleware(httpMiddlewareOptions)
            // .withLoggerMiddleware()
            .build();
    }
    return ctClient;
};

const getApiRoot = (): ByProjectKeyRequestBuilder => {
    return createApiBuilderFromCtpClient(createCtClient()).withProjectKey({
        projectKey: process.env.CT_PROJECT_ID || '',
    });
};

export const getCustomerProfile = async (customerId: string): Promise<Customer | undefined> => {
    logger.info(`Getting customer ${customerId} in commercetools`);
    let ctCustomer: Customer;

    try {
        ctCustomer = (await getApiRoot().customers().withId({ ID: customerId }).get().execute()).body;
    } catch (error: any) {
        logger.error(`Error getting customer in CT with id ${customerId}, status: ${error.statusCode}`, error);
        throw new StatusError(
            error.statusCode,
            `CT get customer api returns failed with status code ${error.statusCode}, msg: ${error.message}`,
        );
    }

    return ctCustomer;
};

export const getOrderById = async (orderId: string): Promise<Order | undefined> => {
    logger.info(`Getting order ${orderId} in commercetools`);
    let ctOrder: Order;

    try {
        ctOrder = (await getApiRoot().orders().withId({ ID: orderId }).get().execute()).body;
    } catch (error) {
        console.log(error);
        return undefined;
    }

    return ctOrder;
};
