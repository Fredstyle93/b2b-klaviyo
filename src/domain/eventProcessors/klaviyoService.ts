import { ConfigWrapper, Events } from 'klaviyo-api';
import logger from '../../utils/log';

ConfigWrapper(process.env.KLAVIYO_AUTH_KEY);

export const sendEventToKlaviyo = (event: KlaviyoEvent) => {
    try {
        if (event.body) {
            logger.debug('Sending event to Klaviyo: ', event.body);
            Events.createEvent(event);
        }
    } catch (error) {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        console.error(`An error was thrown check the HTTP code with ${error.status}`);
    }
};
