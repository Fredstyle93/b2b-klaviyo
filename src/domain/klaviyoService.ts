import { Client, ConfigWrapper, Events, Profiles } from 'klaviyo-api';
import logger from '../utils/log';

ConfigWrapper(process.env.KLAVIYO_AUTH_KEY);

export const sendEventToKlaviyo = async (event: KlaviyoEvent) => {
    logger.info('Sending event to Klaviyo', { zdata: event.body });
    switch (event.type) {
        case 'event':
            return Events.createEvent(event.body);
        case 'profileCreated':
            return Profiles.createProfile(event.body);
        case 'profileResourceUpdated':
            return Profiles.updateProfile(event.body, event.body.data?.id);
        case 'profileUpdated':
            return Client.createClientProfile(event.body, process.env.KLAVIYO_COMPANY_ID);
        default:
            throw new Error(`Unsupported event type ${event.type}`);
    }
};

export const getKlaviyoProfileByExternalId = async (externalId: string): Promise<ProfileType | undefined> => {
    logger.info(`Getting profile in Klaviyo with externalId ${externalId}`);
    try {
        const filter = `equals(external_id,"${externalId}")`;
        const profiles = await Profiles.getProfiles({ filter });
        logger.debug('Profiles response', profiles);
        const profile = profiles?.body.data?.find(
            (profile: ProfileType) => profile.attributes.external_id === externalId,
        );
        logger.debug('Profile', profile);
        return profile;
    } catch (e) {
        logger.error(`Error getting profile in Klaviyo with externalId ${externalId}`);
        throw e;
    }
};
