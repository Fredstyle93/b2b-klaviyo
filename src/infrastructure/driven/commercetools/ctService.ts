import {
    AuthMiddlewareOptions,
    Client,
    ClientBuilder,
    QueueMiddlewareOptions,
    HttpMiddlewareOptions,
} from '@commercetools/sdk-client-v2';
import { ApiRoot, Payment, createApiBuilderFromCtpClient, Customer, Order } from '@commercetools/platform-sdk'
import logger from '../../../utils/log';
import fetch from 'node-fetch';
import { StatusError } from '../../../types/errors/StatusError';

type ByProjectKeyRequestBuilder = ReturnType<ApiRoot['withProjectKey']>;

let ctClient: Client;

export const getCtClient = (concurrency = 10): Client => {
    if (!ctClient) {
        const ctApiCredentials = JSON.parse(process.env.CT_API_CLIENT || '{"clientId":"","secret":""}');

        if(!process.env.CT_AUTH_URL){
            throw Error("Commercetools Host URL not set in environment variable CT_AUTH_URL. Set the variable and restart the application")
        }
        if(!process.env.CT_PROJECT_ID){
            throw Error("Commercetools Project ID not set in environment variable CT_PROJECT_ID. Set the variable and restart the application")
        }
        if(!process.env.CT_API_URL){
            throw Error("Commercetools API URL not set in environment variable CT_API_URL. Set the variable and restart the application")
        }
        const authMiddlewareOptions: AuthMiddlewareOptions = {
            host: process.env.CT_AUTH_URL,
            projectKey: process.env.CT_PROJECT_ID,
            credentials: {
                clientId: ctApiCredentials.clientId,
                clientSecret: ctApiCredentials.secret,
            },
            scopes: [process.env.CT_SCOPES || ''],
            fetch,
        };

        const httpMiddlewareOptions: HttpMiddlewareOptions = {
            maskSensitiveHeaderData: true,
            host: process.env.CT_API_URL,
            enableRetry: true,
            fetch,
        };

        const queueMiddlewareOptions: QueueMiddlewareOptions = {
            concurrency,
        };

        ctClient = new ClientBuilder()
            .withClientCredentialsFlow(authMiddlewareOptions)
            .withHttpMiddleware(httpMiddlewareOptions)
            .withQueueMiddleware(queueMiddlewareOptions)
            // .withLoggerMiddleware()
            .build();
    }
    return ctClient;
};

export const getApiRoot = (): ByProjectKeyRequestBuilder => {
    return createApiBuilderFromCtpClient(getCtClient()).withProjectKey({
        projectKey: process.env.CT_PROJECT_ID || '',
    });
};

export const getCustomerProfile = async (customerId: string): Promise<Customer> => {
    logger.info(`Getting customer ${customerId} in commercetools`);

    try {
        return (await getApiRoot().customers().withId({ ID: customerId }).get().execute()).body;
    } catch (error: any) {
        logger.error(`Error getting customer in CT with id ${customerId}, status: ${error.statusCode}`, error);
        throw new StatusError(
            error.statusCode,
            `CT get customer api returns failed with status code ${error.statusCode}, msg: ${error.message}`,
        );
    }
};

//todo move to ctOrderService
export const getOrderById = async (orderId: string): Promise<Order | undefined> => {
    logger.info(`Getting order ${orderId} in commercetools`);
    let ctOrder: Order;

    try {
        ctOrder = (await getApiRoot().orders().withId({ ID: orderId }).get().execute()).body;
    } catch (error) {
        logger.error(error);
        return undefined;
    }

    return ctOrder;
};

//todo move to ctOrderService
export const getOrderByPaymentId = async (paymentId: string): Promise<Order> => {
    logger.info(`Getting order with payment ${paymentId} in commercetools`);

    try {
        const orderResults = (await getApiRoot().orders().get({
            queryArgs: {
                limit: 1,
                where: `paymentInfo(payments(id = "${paymentId}"))`,
                expand: 'paymentInfo.payments[*]',
            },
        }).execute()).body?.results;
        if (!orderResults?.length) {
            throw new StatusError(
                404,
                `No results returned when querying orders with payment ID ${paymentId}`,
            );
        }
        return orderResults[0];
    } catch (error: any) {
        logger.error(`Error getting order in CT with payment id ${paymentId}, status: ${error.status || error.statusCode}`, error);
        throw new StatusError(
            error.statusCode,
            `CT get order by payment id api returns failed with status code ${error.status || error.statusCode}, msg: ${error.message}`,
        );
    }
};

//todo move to ctPaymentService
export const getPaymentById = async (paymentId: string): Promise<Payment> => {
    logger.info(`Getting payment ${paymentId} in commercetools`);
    let ctPayment: Payment;

    try {
        ctPayment = (await getApiRoot().payments().withId({ ID: paymentId }).get().execute()).body;
    } catch (error: any) {
        logger.error(`Error getting payment in CT with id ${paymentId}, status: ${error.statusCode}`, error);
        throw new StatusError(
            error.statusCode,
            `CT get payment api returns failed with status code ${error.statusCode}, msg: ${error.message}`,
        );
    }

    return ctPayment;
};
