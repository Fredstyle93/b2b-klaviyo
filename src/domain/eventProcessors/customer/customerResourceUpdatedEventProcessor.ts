import { AbstractEvent } from '../abstractEvent';
import logger from '../../../utils/log';
import { getCustomerProfile } from '../../ctService';
import { ResourceUpdatedDeliveryPayload } from '@commercetools/platform-sdk';
import { getKlaviyoProfileByExternalId } from '../../klaviyoService';
import { mapCTCustomerToKlaviyoProfile } from './mappers/CTCustomerToKlaviyoProfileMapper';
import { StatusError } from '../../../types/errors/StatusError';

export class CustomerResourceUpdatedEventProcessor extends AbstractEvent {
    isEventValid(): boolean {
        return this.ctMessage.notificationType === 'ResourceUpdated' && this.ctMessage.resource.typeId === 'customer';
    }

    async generateKlaviyoEvents(): Promise<KlaviyoEvent[]> {
        const message = this.ctMessage as unknown as ResourceUpdatedDeliveryPayload;
        logger.info(`processing CT ${message.resource.typeId}${message.notificationType} message`);
        const customer = await getCustomerProfile(message.resource.id);
        if (!customer) {
            return [];
        }
        const klaviyoProfile = await getKlaviyoProfileByExternalId(message.resource.id);
        if (!klaviyoProfile || !klaviyoProfile.id) {
            throw new StatusError(404, `Profile not found in klaviyo with external id ${message.resource.id}`);
        }

        const body: ProfileRequest = {
            data: {
                type: 'profile',
                id: klaviyoProfile?.id,
                attributes: mapCTCustomerToKlaviyoProfile(customer),
            },
        };
        return [
            {
                body: body,
                type: 'profileResourceUpdated',
            },
        ];
    }
}
