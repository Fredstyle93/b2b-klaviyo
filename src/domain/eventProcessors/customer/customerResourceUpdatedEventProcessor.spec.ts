import { MessageDeliveryPayload } from '@commercetools/platform-sdk/dist/declarations/src/generated/models/subscription';
import { mockDeep } from 'jest-mock-extended';
import { expect as exp } from 'chai';
import { CustomerResourceUpdatedEventProcessor } from './customerResourceUpdatedEventProcessor';
import { getSampleCustomerResourceUpdatedMessage } from '../../../test/testData/ctCustomerMessages';
import { getCustomerProfile } from '../../ctService';
import { getKlaviyoProfileByExternalId } from '../../klaviyoService';
import { getSampleCustomerApiResponse } from '../../../test/testData/ctCustomerApi';
import mocked = jest.mocked;

jest.mock('../../ctService');
jest.mock('../../klaviyoService');
describe('CustomerResourceUpdatedEventProcessor > isEventValid', () => {
    it('should return valid when the customer event has all the required fields', async () => {
        const ctMessageMock: MessageDeliveryPayload = mockDeep<MessageDeliveryPayload>();
        Object.defineProperty(ctMessageMock, 'resource', { value: { typeId: 'customer' } }); //mock readonly property
        Object.defineProperty(ctMessageMock, 'notificationType', { value: 'ResourceUpdated' });

        const event = CustomerResourceUpdatedEventProcessor.instance(ctMessageMock);

        exp(event.isEventValid()).to.be.true;
    });

    it.each`
        resource      | notificationType
        ${'invalid'}  | ${'ResourceUpdated'}
        ${'customer'} | ${'invalid'}
    `(
        'should return invalid when is not a valid customer ResourceUpdated message',
        ({ resource, notificationType }) => {
            const ctMessageMock: MessageDeliveryPayload = mockDeep<MessageDeliveryPayload>();
            Object.defineProperty(ctMessageMock, 'resource', { value: { typeId: resource } });
            Object.defineProperty(ctMessageMock, 'notificationType', { value: notificationType });

            const event = CustomerResourceUpdatedEventProcessor.instance(ctMessageMock);

            exp(event.isEventValid()).to.be.false;
        },
    );
});

describe('CustomerResourceUpdatedEventProcessor > generateKlaviyoEvent', () => {
    it('should generate the klaviyo event when the input customer ResourceUpdated event is valid', async () => {
        const inputMessage = getSampleCustomerResourceUpdatedMessage();
        const message = inputMessage as unknown as MessageDeliveryPayload;
        const event = CustomerResourceUpdatedEventProcessor.instance(message);
        const getCustomerProfileMock = mocked(getCustomerProfile);
        getCustomerProfileMock.mockResolvedValue(getSampleCustomerApiResponse());
        const getKlaviyoProfileMock = mocked(getKlaviyoProfileByExternalId);
        getKlaviyoProfileMock.mockResolvedValue({ type: 'profile', id: 'someId', attributes: {} });

        const klaviyoEvent = await event.generateKlaviyoEvents();

        exp(klaviyoEvent).to.not.be.undefined;
        exp(klaviyoEvent.length).to.be.eq(1);
        expect(klaviyoEvent[0].body).toMatchSnapshot();
    });
});
