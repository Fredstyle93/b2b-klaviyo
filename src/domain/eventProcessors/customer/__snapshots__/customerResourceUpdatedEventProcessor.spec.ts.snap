// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerResourceUpdatedEventProcessor > generateKlaviyoEvent should generate the klaviyo event when the input customer ResourceUpdated event is valid 1`] = `
{
  "data": {
    "attributes": {
      "email": "Alfred_Lebsack@hotmail.com",
      "first_name": "Jailyn",
      "last_name": "Hilll",
      "location": {
        "address1": "12, First Street",
        "address2": "",
        "city": "Brookefort",
        "country": "NL",
        "region": undefined,
        "zip": "12345",
      },
      "organization": undefined,
      "phone_number": "+393234567892",
      "title": "Miss",
    },
    "id": "someId",
    "type": "profile",
  },
}
`;
